source/agora/test/Base.d

```diff
@@ -974,108 +972,61 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
             banman : ban_conf,
             node : self,
-            network : test_conf.configure_network ? [prev_node.address] : null,
+            network : assumeUnique(network),
         };
 
         return conf;
     }
 
+    auto keys = WK.Keys.byRange();
+    auto nextKey () { scope (exit) keys.popFront(); return keys.front; }
     NodeConfig[] node_configs;
     Config[] configs;
 
-    final switch (test_conf.topology)
-    {
-    case NetworkTopology.Simple:
-        node_configs = iota(test_conf.nodes).map!(idx => makeNodeConfig(true, WK.Keys[idx])).array;
-        configs = iota(test_conf.nodes)
-            .map!(idx => makeConfig(node_configs[idx], node_configs)).array;
-        break;
-
-    case NetworkTopology.FindNetwork:
-        node_configs = iota(test_conf.nodes).map!(idx => makeNodeConfig(false, WK.Keys[idx])).array;
-        configs = iota(test_conf.nodes)
-            .map!(idx => makeMinimalNetwork(idx, node_configs[idx], node_configs)).array;
-        break;
-
-    case NetworkTopology.FindQuorums:
-        node_configs = iota(test_conf.nodes).map!(idx => makeNodeConfig(true, WK.Keys[idx])).array;
-        configs = iota(test_conf.nodes)
-            .map!(idx => makeMinimalNetQuorum(idx, node_configs[idx], node_configs)).array;
-        break;
-
-    case NetworkTopology.OneNonValidator:
-        node_configs ~= iota(test_conf.nodes - 1).map!(idx => makeNodeConfig(true, WK.Keys[idx])).array;
-        node_configs ~= makeNodeConfig(false, WK.Keys[node_configs.length]);
-        configs = iota(test_conf.nodes)
-            .map!(idx => makeConfig(node_configs[idx], node_configs)).array;
-        break;
-
-    case NetworkTopology.OneValidator:
-        node_configs ~= makeNodeConfig(true, WK.Keys[0]);
-        node_configs ~= iota(test_conf.nodes - 1).map!(idx => makeNodeConfig(false, WK.Keys[idx + 1])).array;
-        configs = iota(test_conf.nodes)
-            .map!(idx => makeConfig(node_configs[idx], node_configs)).array;
-        break;
-
-    case NetworkTopology.OneFullNodeOutsider:
-        node_configs ~= iota(test_conf.nodes).map!(idx => makeNodeConfig(true, WK.Keys[idx])).array;
-        configs = iota(test_conf.nodes)
-            .map!(idx => makeConfig(node_configs[idx], node_configs)).array;
-
-        // add one non-validator outside the network
-        node_configs ~= makeNodeConfig(false, WK.Keys[node_configs.length]);
-        configs ~= makeConfig(node_configs[$ - 1], node_configs);
-        break;
-
-    case NetworkTopology.TwoOutsiderValidators:
-        node_configs ~= iota(test_conf.nodes).map!(idx => makeNodeConfig(true, WK.Keys[idx])).array;
-        configs = iota(test_conf.nodes)
-            .map!(idx => makeConfig(node_configs[idx], node_configs)).array;
-        break;
-    }
+    test_conf.validators.iota
+        .each!(_ => node_configs ~= makeNodeConfig(true, nextKey()));
+    test_conf.full_nodes.iota
+        .each!(_ => node_configs ~= makeNodeConfig(false, nextKey()));
```

Comment:
That looks like a prime use case for `zip`.

Reply:
Or `refRange`

Reply:
But refrange doesn't change the semantics..

Reply:
Or I just don't know how to use it properly?

```D
import std.range;
import std.stdio;

void main ()
{
    auto keys = iota(5);

    auto r = refRange(&keys);

    foreach (x; r)
        writeln(x);
    foreach (x; r)
        writeln(x);
}
```

Prints the same range twice.

Reply:
```D
auto keys = WK.Keys.byRange().refRange();
keys.take(test_conf.validators).each!(key => node_configs ~= makeNodeConfig(true, key));
keys.take(test_conf.full_nodes).each!(key => node_configs ~= makeNodeConfig(false, key));
```

Would this work ?

Reply:
First of all refRange takes a pointer.

Reply:
(don't shoot the messenger)

Reply:
You got bamboozled by `foreach`. I have a bug report open for this.
```D
import std.range;
import std.stdio;
import std.algorithm;

void main ()
{
    auto keys = iota(5);

    auto r = refRange(&keys);
    r.each!writeln;
    r.each!writeln;
}
```

FTFY

Reply:
> /Users/andrejmitrovic/Downloads/ldc2-1.22.0-osx-x86_64/bin/../import/std/range/package.d(11502,13): Error: cannot modify struct instance *this._range of type Range because it contains const or immutable members

haha ranges go compiler errrrrror

Reply:
I'm reducing it.

Reply:
Reduced test-case:

```D
import std.range;

public static auto byRange ()
{
    static struct Range
    {
        size_t lbound = 0;

        // trigger point (remove const and it compiles)
        const size_t hbound = 100;

        size_t length () const { return 0; }
        bool empty () const { return false; }
        int front () const { return 0; }
        void popFront () { }
    }

    return Range(0);
}

void main ()
{
    auto keys = byRange();
    auto r = refRange(&keys);
}
```

Time to submit it..

Reply:
we can also work around it by making `hbound` an enum